src/mlpack/methods/ann/loss_functions/multilabel_softmargin_loss.hpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * @file multilabel_softmargin_loss.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Definition of the Multi Label Soft Margin Loss function.
+ * It is a criterion that optimizes a multi-label one-versus-all loss based on
+ * max-entropy, between input x and target y of size (N, C) where N is the
+ * batch size and C is the number of classes.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_HPP
+#define MLPACK_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class MultiLabelSoftMarginLoss
+{
+ public:
+  /**
+   * Create the MultiLabelSoftMarginLoss object.
+   *
+   * @param weight A manual rescaling weight given to each class. Weight is a
+   *               (1, C) vector where C is the number of classes.
+   * @param num_classes The number of classes, C.
+   * @param reduction Specifies the reduction to apply to the output. If false,
+   *                  'mean' reduction is used, where sum of the output will be
+   *                  divided by the number of elements in the output. If
+   *                  true, 'sum' reduction is used and the output will be
+   *                  summed. It is set to true by default.
+   */
+  MultiLabelSoftMarginLoss(arma::mat weight,
```

Comment:
Do we need to have the weight matrix here?
The definition of the loss function doesn't state its use. I know pytorch takes it as an input (by default it is none). However in your implementation this is a required parameter (not sure why). I might have missed something. Let me know what you think.

Reply:
In the PyTorch implementation, default value of weight is None. In case the default is used, the weights matrix is initialized with all 1's and otherwise, the specified values for each class will be used. In my implementation here, I simply feel that keeping the weight matrix as  compulsory input in the constructor offers the end user more flexibility than having to access it via accessor methods. Also, this allows the user to be more conscious when choosing the weights for each class. They can choose it to be all 1's or they can assign separate weights for each class, but whatever they do they will focus on it once only when they are initializing the loss function and won't need to bother about it after that. This is why I have gone for this approach. @kartikdutt18 

Coming to the definition of the function, some definitions include the weight matrix and some don't. The PyTorch implementation includes it and so I have included it.

Reply:
Since C++ doesn't really have None type for objects (not pointers), So here the constructor can't be called without passing the weight matrix. The weights matrix is used when there is class imbalance and I think most people won't be passing weights matrix as input as it might not gel very well with the API. If you want to keep the weights matrix, I would personally prefer setting it to a default matrix. Let me know what you think.

Reply:
I would like to keep it. How would I do that without following the current implementation? Because, I need the weight matrix to be of the size (1,numClasses) and hence I can't just set it to a default value within the constructor parameter list itself, can I?

Reply:
So what I would do incase I had to use a weight matrix then I would have a private weight matrix with setters and getters. Constructor will initialise the matrix to default value of 1. User can then set weights. This is similar to weights for other layers. Having an unnecessary matrix in constructor will be a hinderance to most users. Let me know what you think.

Reply:
> The reason I felt weight matrix as input is redundant because of this. User shouldn't set this to 1 if he doesn't want to change weight for different classes.

@sreenikSS, what do you think about this?

Reply:
Yep. I have thought of this too. In fact, this was the implementation I was going for initially, before I thought of going with the current implementation. I have explained my reasoning in a response to your comment below.  

The main idea I am going for is that the users should be consciously choosing the weight values irrespective of whether they are using the default values or not, when they are using a specific loss function like this where the weights can be specified for each class. For some reason that appeals to me more somehow. I would like some more opinion on this from others as well. Changing the implementation shouldn't take much time, compared to gathering opinions :)

Reply:
Hmm, **Let's wait then**. It's just that we can add this to multiple loss functions here and if that was to be done they shouldn't hinder with existing loss functions implementation. Also I think no other framework (Keras, PyTorch, Tensorflow) asks users to set the weight to 1. Mainly because most users will prefer if they wrote less code on their side.

Reply:
@kartikdutt18 I think I have a common solution to your problem, which can take into account both your idea and mine! Let's have 2 constructors for the loss function. The default constructor is the one which most users will be using and that will set the weights to 1 without the users having to write anything extra. And then we can have the overloaded constructor, which will follow the current implementation and require the user to specify the weights. I think this is going to be a much more neat idea than having to use accessor functions to modify the weight vector(which I am not sure as of yet is possible or not, explained below)! Let me know what you think of this. Thanks!

Also, I had a doubt. Assume that we wanted to use accessor methods. Once the loss function is instantiated using the constructor, and Added to the model, where exactly will the user be able to use a setter method to modify the weight. Is that even possible? Like, how exactly will I be modifying the Forward function in that case, if at all that is possible? Would be really helpful if you can provide an example so that I can visualize it better.

Reply:
> Also, I had a doubt. Assume that we wanted to use accessor methods. Once the loss function is instantiated using the constructor, and Added to the model, where exactly will the user be able to use a setter method to modify the weight. Is that even possible? Like, how exactly will I be modifying the Forward function in that case, if at all that is possible? Would be really helpful if you can provide an example so that I can visualize it better.

I can clarify on this. Maybe you could refer to implementation of any layer. I would suggest going through the tests for transposed convolution. I think that should help. Let me know if I need to clarify something there.

For the overloaded constructor, that is also a possible solution.

Reply:
Or there is  a PR open for depthwise convolution maybe you could refer to tests for that as well.

Reply:
Sure! I will go through the code for the transposed convolution files and then let you know whether I decide to go ahead with the overloaded constructor approach or the the accessor methods approach. Either way, I should be done with this by tomorrow.

Reply:
I decided to go ahead with the overloaded constructor approach.

Reply:
@kartikdutt18 the usability issue you have pointed out looks reasonable to me. Even though the specialty of this loss is allowing the user to set weights as what @iamshnoo has pointed out, I don't think forcing the user to pass a vector would be a good idea.

However, the idea of overloading constructors looks good to me.

Reply:
Ohk, I think I will show the code of what I thought in the evening today. I don't mind going for overloaded operator but I think the way, layers are implemented might be a better approach.